standalone/api.c

```diff
@@ -782,3 +786,61 @@ void modsecReportRemoteLoadedRules()
 
 }
 #endif
+
+#ifdef WAF_JSON_LOGGING_ENABLE
+void modsecReopenLogfileIfNeeded(request_rec *r)
+{
+    modsec_rec *msr = NULL;
+    int rc = 0;
+    apr_file_t * fd = NULL;
+    /* Find the transaction context first. */
+    msr = retrieve_msr(r);
+
+    if (msr == NULL)
+        return;
+
+    if (msc_waf_log_reopened){
+        if (msr->modsecurity != NULL && msr->modsecurity->wafjsonlog_lock != NULL){
+            rc = waf_get_exclusive_lock(msr->modsecurity->wafjsonlog_lock);
+            if (waf_lock_is_error(rc)) {
+                ap_log_rerror(APLOG_MARK, APLOG_ERR | APLOG_NOERRNO, 0, r,
+                        "ModSecurity not able to get lock for %s file", msc_waf_log_path);
+                return;
+            }
+
+            rc = apr_file_open(&fd, msc_waf_log_path,
```

Comment:
do we need to close before reopening?

Reply:
I suggested that we first open the new file (that is, btw, what Nginx also does - http://hg.nginx.org/nginx/file/tip/src/core/ngx_cycle.c#l1155 ) because, if it fails, we at least don't have the old one closed and can keep writing logs into it until the next attempt is taken.
One of the guarantees that WAF strives to provide is that no logs are missing, and if we closed the file first and then failed to reopen it, we would be in a state when we have nowhere to write logs to.